File: openags/research/adapter.py
"""Runtime adapter — converts SOUL.md + skills + memory into CLI agent config files.

Before launching an external CLI agent (Claude Code, Codex, etc.) in a folder,
this module reads the OpenAGS configuration and generates the equivalent config
file that the CLI agent will auto-load.

Adapter mapping:
    Claude Code  → CLAUDE.md (auto-read from cwd)
    Codex        → AGENTS.md
    Copilot      → .github/copilot-instructions.md
    Gemini CLI   → GEMINI.md
"""

from __future__ import annotations

import logging
from pathlib import Path

logger = logging.getLogger(__name__)


def _read_soul(folder: Path) -> str:
    """Read SOUL.md body (the prompt part, not frontmatter)."""
    soul_path = folder / "SOUL.md"
    if not soul_path.exists():
        return ""

    from openags.agent.soul import parse_soul

    _, body = parse_soul(soul_path)
    return body


def _read_skills(folder: Path) -> list[str]:
    """Read all skill .md bodies from folder/skills/."""
    skills_dir = folder / "skills"
    if not skills_dir.is_dir():
        return []

    bodies: list[str] = []
    for path in sorted(skills_dir.rglob("*.md")):
        text = path.read_text(encoding="utf-8")
        # Strip frontmatter if present
        if text.startswith("---"):
            try:
                end = text.index("---", 3)
                text = text[end + 3:].strip()
            except ValueError:
                pass
        if text.strip():
            bodies.append(text)
    return bodies


def _read_memory(folder: Path) -> str:
    """Read memory.md content."""
    memory_path = folder / "memory.md"
    if not memory_path.exists():
        return ""
    content = memory_path.read_text(encoding="utf-8").strip()
    return content


def _read_auto_memory(folder: Path) -> str:
    """Read MEMORY.md (auto-learned knowledge)."""
    mem_path = folder / "MEMORY.md"
    if not mem_path.exists():
        return ""
    content = mem_path.read_text(encoding="utf-8").strip()
    # Only use first 200 lines (same as auto_memory.py convention)
    lines = content.splitlines()
    if len(lines) > 200:
        lines = lines[:200]
    return "\n".join(lines)


def _build_combined_prompt(folder: Path) -> str:
    """Build a combined system prompt from SOUL.md + skills + memory."""
    parts: list[str] = []

    # 1. SOUL.md body (role definition)
    soul = _read_soul(folder)
    if soul:
        parts.append(soul)

    # 2. Skills
    skills = _read_skills(folder)
    if skills:
        parts.append("## Skills\n")
        for skill_body in skills:
            parts.append(skill_body)

    # 3. Memory context
    memory = _read_memory(folder)
    if memory:
        parts.append(f"## Project Memory\n\n{memory}")

    # 4. Auto-learned memory (only if different from memory.md)
    auto_mem = _read_auto_memory(folder)
    if auto_mem and auto_mem != memory:
        parts.append(f"## Learned Knowledge\n\n{auto_mem}")

    return "\n\n".join(parts)


def generate_claude_md(folder: Path) -> Path:
    """Generate CLAUDE.md for Claude Code from SOUL.md + skills + memory.

    Claude Code auto-reads CLAUDE.md from the working directory.
    Returns the path to the generated file.
    """
    content = _build_combined_prompt(folder)
    if not content:
        content = f"You are an AI assistant working in the {folder.name} directory."

    out = folder / "CLAUDE.md"
    out.write_text(content + "\n", encoding="utf-8")
    logger.info("Generated CLAUDE.md in %s (%d chars)", folder, len(content))
    return out


def generate_agents_md(folder: Path) -> Path:
    """Generate AGENTS.md for Codex from SOUL.md + skills + memory."""
    content = _build_combined_prompt(folder)
    if not content:
        content = f"You are an AI assistant working in the {folder.name} directory."

    out = folder / "AGENTS.md"
    out.write_text(content + "\n", encoding="utf-8")
    logger.info("Generated AGENTS.md in %s (%d chars)", folder, len(content))
    return out


def generate_copilot_md(folder: Path) -> Path:
    """Generate .github/copilot-instructions.md for Copilot."""
    content = _build_combined_prompt(folder)
    if not content:
        content = f"You are an AI assistant working in the {folder.name} directory."

    out_dir = folder / ".github"
    out_dir.mkdir(exist_ok=True)
    out = out_dir / "copilot-instructions.md"
    out.write_text(content + "\n", encoding="utf-8")
    logger.info("Generated copilot-instructions.md in %s", folder)
    return out


def generate_gemini_md(folder: Path) -> Path:
    """Generate GEMINI.md for Gemini CLI."""
    content = _build_combined_prompt(folder)
    if not content:
        content = f"You are an AI assistant working in the {folder.name} directory."

    out = folder / "GEMINI.md"
    out.write_text(content + "\n", encoding="utf-8")
    logger.info("Generated GEMINI.md in %s", folder)
    return out


# Registry: runtime type → generator function
_GENERATORS: dict[str, object] = {
    "claude_code": generate_claude_md,
    "codex": generate_agents_md,
    "copilot": generate_copilot_md,
    "gemini_cli": generate_gemini_md,
}


def prepare_folder_for_cli(folder: Path, runtime_type: str) -> Path | None:
    """Prepare a folder for a specific CLI agent runtime.

    Reads SOUL.md + skills + memory and generates the appropriate config file
    that the CLI agent will auto-load when started in this folder.

    Returns the generated file path, or None if no generator exists.
    """
    generator = _GENERATORS.get(runtime_type)
    if generator is None:
        logger.warning("No adapter for runtime '%s'", runtime_type)
        return None
    return generator(folder)  # type: ignore[operator]

File: packages/desktop/src/main/providers/adapter.ts
/**
 * Adapter — converts SOUL.md + skills + memory into CLI agent config files.
 *
 * Before sending a message to Claude Code / Codex / Gemini, this reads the
 * OpenAGS folder structure and generates the config file the CLI agent auto-loads.
 *
 * Mapping:
 *   Claude Code → CLAUDE.md
 *   Codex       → AGENTS.md
 *   Gemini CLI  → GEMINI.md
 *   Cursor      → CLAUDE.md (same as Claude)
 */

import * as fs from 'fs'
import * as path from 'path'

/** Read SOUL.md body (strip YAML frontmatter, keep the prompt). */
function readSoulBody(folder: string): string {
  const soulPath = path.join(folder, 'SOUL.md')
  if (!fs.existsSync(soulPath)) return ''

  const text = fs.readFileSync(soulPath, 'utf-8')
  // Strip frontmatter
  if (text.startsWith('---')) {
    const end = text.indexOf('---', 3)
    if (end !== -1) return text.slice(end + 3).trim()
  }
  return text.trim()
}

/** Read all skill .md files from folder/skills/ (body only, strip frontmatter). */
function readSkills(folder: string): string[] {
  const skillsDir = path.join(folder, 'skills')
  if (!fs.existsSync(skillsDir) || !fs.statSync(skillsDir).isDirectory()) return []

  const bodies: string[] = []
  for (const file of fs.readdirSync(skillsDir).sort()) {
    if (!file.endsWith('.md')) continue
    let text = fs.readFileSync(path.join(skillsDir, file), 'utf-8')
    if (text.startsWith('---')) {
      const end = text.indexOf('---', 3)
      if (end !== -1) text = text.slice(end + 3)
    }
    text = text.trim()
    if (text) bodies.push(text)
  }
  return bodies
}

/** Read memory.md content. */
function readMemory(folder: string): string {
  const memPath = path.join(folder, 'memory.md')
  if (!fs.existsSync(memPath)) return ''
  return fs.readFileSync(memPath, 'utf-8').trim()
}

/** Read MEMORY.md (auto-learned, max 200 lines). */
function readAutoMemory(folder: string): string {
  const memPath = path.join(folder, 'MEMORY.md')
  if (!fs.existsSync(memPath)) return ''
  const lines = fs.readFileSync(memPath, 'utf-8').trim().split('\n')
  return lines.slice(0, 200).join('\n')
}

/** Build combined prompt from SOUL.md + skills + memory. */
function buildPrompt(folder: string): string {
  const parts: string[] = []

  const soul = readSoulBody(folder)
  if (soul) parts.push(soul)

  const skills = readSkills(folder)
  if (skills.length > 0) {
    parts.push('## Skills\n')
    parts.push(...skills)
  }

  const memory = readMemory(folder)
  if (memory) parts.push(`## Project Memory\n\n${memory}`)

  const autoMem = readAutoMemory(folder)
  if (autoMem && autoMem !== memory) parts.push(`## Learned Knowledge\n\n${autoMem}`)

  return parts.join('\n\n')
}

/** All config files that should stay in sync. */
const CONFIG_FILES: Record<string, string> = {
  'SOUL.md': 'soul',
  'CLAUDE.md': 'claude_code',
  'AGENTS.md': 'codex',
  'GEMINI.md': 'gemini_cli',
}

/**
 * Sync all config files in a folder.
 * Finds the most recently modified one, uses it as source, updates the rest.
 * If SOUL.md is the source → extract body (strip frontmatter) for others.
 * If CLAUDE.md/AGENTS.md/GEMINI.md is the source → update SOUL.md body (keep frontmatter).
 */
export function syncConfigFiles(folder: string): void {
  // Find which config file is newest
  let newestFile = ''
  let newestMtime = 0

  for (const filename of Object.keys(CONFIG_FILES)) {
    const p = path.join(folder, filename)
    try {
      const mtime = fs.statSync(p).mtimeMs
      if (mtime > newestMtime) {
        newestMtime = mtime
        newestFile = filename
      }
    } catch { /* doesn't exist */ }
  }

  if (!newestFile) {
    // No config files exist — nothing to sync
    return
  }

  if (newestFile === 'SOUL.md') {
    // SOUL.md is the source → generate others from it (+ skills + memory)
    const content = buildPrompt(folder)
      || `You are an AI assistant working in the ${path.basename(folder)} directory.`

    for (const filename of Object.keys(CONFIG_FILES)) {
      if (filename === 'SOUL.md') continue
      fs.writeFileSync(path.join(folder, filename), content + '\n', 'utf-8')
    }
  } else {
    // A CLI config file is newest → use its content to update all others
    const sourcePath = path.join(folder, newestFile)
    const content = fs.readFileSync(sourcePath, 'utf-8').trim()

    // Update other CLI config files
    for (const filename of Object.keys(CONFIG_FILES)) {
      if (filename === newestFile || filename === 'SOUL.md') continue
      fs.writeFileSync(path.join(folder, filename), content + '\n', 'utf-8')
    }

    // Update SOUL.md body (keep frontmatter)
    const soulPath = path.join(folder, 'SOUL.md')
    if (fs.existsSync(soulPath)) {
      const soulText = fs.readFileSync(soulPath, 'utf-8')
      if (soulText.startsWith('---')) {
        const end = soulText.indexOf('---', 3)
        if (end !== -1) {
          const frontmatter = soulText.slice(0, end + 3)
          fs.writeFileSync(soulPath, frontmatter + '\n\n' + content + '\n', 'utf-8')
        }
      }
    }
  }
}

/**
 * Sync all config files + skill symlinks across an entire project.
 */
export function syncProjectConfigs(projectDir: string): void {
  // Sync module config files (not root — root CLAUDE.md is project-level)
  try {
    for (const entry of fs.readdirSync(projectDir, { withFileTypes: true })) {
      if (!entry.isDirectory() || entry.name.startsWith('.')) continue
      const sub = path.join(projectDir, entry.name)
      const hasSoul = fs.existsSync(path.join(sub, 'SOUL.md'))
      const hasClaude = fs.existsSync(path.join(sub, 'CLAUDE.md'))
      if (hasSoul || hasClaude) {
        syncConfigFiles(sub)
      }
    }
  } catch { /* ignore */ }

  // Sync skill symlinks for Claude Code discovery
  syncSkillSymlinks(projectDir)
}

/**
 * Create .claude/skills/ symlinks so Claude Code can discover our skills.
 * Links project-level skills and module-level skills.
 */
function syncSkillSymlinks(projectDir: string): void {
  // Project-level skills: skills/ → .claude/skills/
  const projectSkills = path.join(projectDir, 'skills')
  if (fs.existsSync(projectSkills) && fs.statSync(projectSkills).isDirectory()) {
    const claudeSkills = path.join(projectDir, '.claude', 'skills')
    fs.mkdirSync(claudeSkills, { recursive: true })

    for (const entry of fs.readdirSync(projectSkills, { withFileTypes: true })) {
      if (!entry.isDirectory()) continue
      const skillDir = path.join(projectSkills, entry.name)
      if (!fs.existsSync(path.join(skillDir, 'SKILL.md'))) continue

      const link = path.join(claudeSkills, entry.name)
      if (!fs.existsSync(link)) {
        try { fs.symlinkSync(skillDir, link) } catch { /* ignore */ }
      }
    }
  }

  // Module-level skills: module/skills/ → module/.claude/skills/
  try {
    for (const entry of fs.readdirSync(projectDir, { withFileTypes: true })) {
      if (!entry.isDirectory() || entry.name.startsWith('.')) continue
      const modDir = path.join(projectDir, entry.name)
      const modSkills = path.join(modDir, 'skills')

      if (!fs.existsSync(modSkills) || !fs.statSync(modSkills).isDirectory()) continue

      const claudeModSkills = path.join(modDir, '.claude', 'skills')
      fs.mkdirSync(claudeModSkills, { recursive: true })

      for (const skill of fs.readdirSync(modSkills, { withFileTypes: true })) {
        if (!skill.isDirectory()) continue
        const skillDir = path.join(modSkills, skill.name)
        if (!fs.existsSync(path.join(skillDir, 'SKILL.md'))) continue

        const link = path.join(claudeModSkills, skill.name)
        if (!fs.existsSync(link)) {
          try { fs.symlinkSync(skillDir, link) } catch { /* ignore */ }
        }
      }
    }
  } catch { /* ignore */ }
}
